api/v1alpha1/rabbitmqsecretenginerole_types.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"reflect"
+
+	vaultutils "github.com/redhat-cop/vault-config-operator/api/v1alpha1/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RabbitMQSecretEngineRoleSpec defines the desired state of RabbitMQSecretEngineRole
+type RabbitMQSecretEngineRoleSpec struct {
+	// Authentication is the k8s auth configuration to be used to execute this request
+	// +kubebuilder:validation:Required
+	Authentication KubeAuthConfiguration `json:"authentication"`
+
+	// Path at which to make the configuration.
+	// The final path will be {[spec.authentication.namespace]}/{spec.path}/config/{metadata.name}.
+	// The authentication role must have the following capabilities = [ "create", "read", "update", "delete"] on that path.
+	// +kubebuilder:validation:Required
+	Path Path `json:"path"`
+
+	// +kubebuilder:validation:Required
+	RMQSERole `json:",inline"`
+}
+
+type RMQSERole struct {
+	// Comma-separated RabbitMQ permissions tags to associate with the user. This determines the level of
+	// access to the RabbitMQ management UI granted to the user. Omitting this field will
+	// lead to a user than can still connect to the cluster through messaging protocols,
+	// but cannot perform any management actions.
+	// +kubebuilder:validation:Optional
+	Tags string `json:"tags,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	Vhosts []Vhost `json:"vhosts,omitempty"`
```

Comment:
when array are used see if any of the processing directive is applicable: https://book.kubebuilder.io/reference/markers/crd-processing.html
this comment referes also to the next field.

Reply:
Added for arrays:
	// +listType=map
	// +listMapKey=vhostName
	// +listMapKey=permissions
	// +kubebuilder:validation:UniqueItems=true

Reply:
Faced an issue with list type map, it requires default value and as this arrays can be empty it's not acceptable. UniqueItems cannot be set to true since the runtime complexity becomes quadratic.
Added // +listType=atomic instead.